### PR TITLE
GH-36120: [C#] Support schema metadata through the C API

### LINKED ISSUE #36120: [C#] Support schema and field metadata in the C Data Interface
### Describe the enhancement requested

.NET support for import and export of schemas over the C data interface was added in #33856, but metadata in fields and schemas is currently ignored.

I'm happy to take a look at this unless this is something @wjones127 or @CurtHagenlocher was already planning on doing?

### Component(s)

C#

### PR DESCRIPTION
**Pull request by CurtHagenlocher**
### What changes are included in this PR?

Import and export of field- and schema-level metadata via the C API.

### Are these changes tested?

Yes

* Closes: #36120

### CODE REVIEW

**Comment by CurtHagenlocher** [comment]
@wjones127 has offered to review ;)

File: csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs [hunk 1]
```diff
@@ -281,21 +283,53 @@ public Field GetAsField()
...
+            private unsafe static IReadOnlyDictionary<string, string> GetMetadata(byte* metadata)
+            {
+                if (metadata == null)
+                {
+                    return null;
+                }
+
+                IntPtr ptr = (IntPtr)metadata;
+                int count = Marshal.ReadInt32(ptr);
+                if (count == 0)
```
Comment (wjones127) [inline csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs:311]: Perhaps, to be somewhat defensive:

```suggestion
                if (count <= 0)
```

File: csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs [hunk 2]
```diff
@@ -281,21 +283,53 @@ public Field GetAsField()
...
+
+                Dictionary<string, string> result = new Dictionary<string, string>(count);
+                for (int i = 0; i < count; i++)
+                {
+                    result[ReadString(ref ptr)] = ReadString(ref ptr);
+                }
+                return result;
+            }
+
+            private unsafe static string ReadString(ref IntPtr ptr)
```
Comment (wjones127) [inline csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs:325]: Might want to name more specifically here:

```suggestion
            private unsafe static string ReadMetadataString(ref IntPtr ptr)
```

File: csharp/test/Apache.Arrow.Tests/CDataInterfacePythonTests.cs
Comment (wjones127) [inline csharp/test/Apache.Arrow.Tests/CDataInterfacePythonTests.cs:1]: Could you add a test that involves some multi-byte unicode characters?

File: csharp/src/Apache.Arrow/C/CArrowSchemaExporter.cs [hunk 3]
```diff
@@ -239,6 +240,51 @@ private static long GetFlags(IArrowType datatype, bool nullable = true)
...
+                WriteString(ref ptr, lengths[i++], pair.Key);
+                WriteString(ref ptr, lengths[i++], pair.Value);
+            }
+
+            Debug.Assert((long)(IntPtr)ptr - (long)result == size);
+
+            return (byte*)result;
+        }
+
+        private unsafe static void WriteString(ref byte* ptr, int length, string str)
```
Comment (wjones127) [inline csharp/src/Apache.Arrow/C/CArrowSchemaExporter.cs:277]:
```suggestion
        private unsafe static void WriteMetadataString(ref byte* ptr, int length, string str)
```

File: csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs [hunk 4]
```diff
@@ -281,21 +283,53 @@ public Field GetAsField()
...
+            {
+                int length = Marshal.ReadInt32(ptr);
```
Comment (wjones127) [inline csharp/src/Apache.Arrow/C/CArrowSchemaImporter.cs:327]: Perhaps we should also throw an exception if length is negative? Or will `GetString` already do that?

**Review by wjones127 — CHANGES_REQUESTED** [review]
Looks pretty good. Would mainly like to see a test for multi-byte characters, and then have a few smaller nits you can consider.

**Review by wjones127 — APPROVED** [review]
Looks good now. Thanks!

**Pull request merged**